### PR TITLE
Open id connect

### PR DESCRIPTION
**Pull request by antarcticrainforest**
- These changes untie the relation of the auth mechanism to keycloak
  rather than using exclusively keycloak to API should be able to work
  with any kind of open connect id provider.
- The client uses authlib for authentication.

### CODE REVIEW

**Comment by MoSHad91** [comment]
i u don't mind, on the weekend I will have a look, I'm busy with developing things for regiklim at the moment

**Comment by antarcticrainforest** [comment]
Hmm, I actually need to deploy this on a machine for beta testing. For this I need to change the deployment, ideally, I'd like to work with a version that isn't outdated.

**Comment by antarcticrainforest** [comment]
I am going to push a dev version to the container registry and use this.

File: freva-rest/src/freva_rest/__init__.py
Comment (MoSHad91) [inline freva-rest/src/freva_rest/__init__.py:4]: there is no standard way to make it dynamic to inherits from the `freva-rest` version [here](https://github.com/FREVA-CLINT/freva-nextgen/blob/open-id-connect/freva-client/src/freva_client/__init__.py#L20) or vice versa? generally I'm not fan of defining version in `__init__.py` and instead of defining version in diff files, my suggestion is to define it in `pyproject.toml` and build the image based on the version in pyproject.

File: freva-rest/src/freva_rest/auth.py
Comment (MoSHad91) [inline freva-rest/src/freva_rest/auth.py:60]: instead of parsing the request before fetch_or_refresh_token stage, if we get the whole request as Request here, we don't need to define and take care of so many attrs complicated typing i think:

```python
from fastapi import Request
async def fetch_or_refresh_token(request: Request):
    data = {
        "client_id": request.client_id or server_config.oidc_client,
        "client_secret": request.client_secret or os.getenv("OIDC_CLIENT_SECRET", ""),
        "grant_type": request.grant_type,
        "username": request.username,
        "password": request.password,
        "refresh_token": request.refresh_token,
    }
    data = {k: v for k, v in data.items() if v}
```

tbh I haven't tried in fastapi but it seems it has to be feasible ...

File: freva-rest/src/freva_rest/auth.py
Comment (MoSHad91) [inline freva-rest/src/freva_rest/auth.py:119]: another point, in data-loader we perfectly consider ssl certs to secure the connection but here we don't use it here. But the problem which is gonna happen, i think it would be in `dev`, which we need to consider an env for this:


```python
ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
ssl_context.load_cert_chain('/path/to/cert.pem', keyfile='/path/to/key.pem')
async with aiohttp.ClientSession(timeout=TIMEOUT, connector=aiohttp.TCPConnector(ssl=ssl_context)) as client:
```
which `/path/to/cert.pem` and `/path/to/key.pem` can comes from env. because we use them in main function of cli of dataportal as well ...

https://medium.com/@mariovanrooij/adding-https-to-fastapi-ad5e0f9e084e

File: freva-rest/src/freva_rest/cli.py
Comment (MoSHad91) [inline freva-rest/src/freva_rest/cli.py:165]: beautiful  👍

**Review by MoSHad91 — COMMENTED** [review]
Excelent @antarcticrainforest 

just dropped some minor comments, but after clarification of Token typedict it looks good to me

File: freva-client/src/freva_client/auth.py
Comment (MoSHad91) [reply freva-client/src/freva_client/auth.py:75]: ahhh I got where the problem comes form! In both `freva_rest` and `freva_client` you defined Token with different structure and attrs which I have no clue why we don't define the centralized Token class from the originated location and then use in different endpoints. It made me a bit confused in the first tbh

File: docs/source/auth/index.rst
Comment (antarcticrainforest) [reply docs/source/auth/index.rst:49]: Ok

File: docs/source/auth/index.rst
Comment (antarcticrainforest) [reply docs/source/auth/index.rst:84]: This is the response from keycloak, therefore I added a response as they come from keycloak.

File: freva-client/src/freva_client/auth.py
Comment (antarcticrainforest) [reply freva-client/src/freva_client/auth.py:44]: Is that the case? TBH I would be more pragmatic here and only do something about this when a problem occurs.

File: freva-client/src/freva_client/auth.py
Comment (antarcticrainforest) [reply freva-client/src/freva_client/auth.py:75]: The freva rest code defines the token dat as it comes from the keycloak - I didnt' want to change it. The client lib makes an absolute expiry timestamp from the relative expires in value. We could let the restAPI return a json payload that already defines the absolute token expiry time. But I would like to keep the interference of the restAPI with the auth system as little as possible.

File: freva-client/src/freva_client/auth.py
Comment (antarcticrainforest) [reply freva-client/src/freva_client/auth.py:114]: See my comment above.

File: freva-rest/src/freva_rest/__init__.py
Comment (antarcticrainforest) [reply freva-rest/src/freva_rest/__init__.py:4]: I would like to be able to do this:

```python
import foo
print(foo.__version__)
```

This is important, because when people report issues the first we want to ask is what version they are using. 

I know what you could muck around with pkg_resources or importlib. But I think it's very clunky. Hence I will keep this __version__ constructs.

File: freva-rest/src/freva_rest/auth.py
Comment (antarcticrainforest) [reply freva-rest/src/freva_rest/auth.py:60]: Yes, that was indeed my first approach. But then I decided to follow this approach because it integrates much better with the swagger help.

File: freva-rest/src/freva_rest/auth.py
Comment (antarcticrainforest) [reply freva-rest/src/freva_rest/auth.py:119]: I am not sure I understand this. The data-loader isn't involved here.

The ssl connection to the keycloak shouldn't be based on a self-signed cert. 

On the other hand for the redis connection in the data-loader, we want to use a self-signed cert to make sure that only hosts that hold the private key can make connections to the redis host.

**Review by antarcticrainforest — COMMENTED** [review]
Ok thanks for looking at this.

The Question about the Token. I think the `expires_in` data that comes from keycloak is pretty useless as we have to know when the token was issued. But better would be a absolute `expires_at` value. We could already convert the `expires_in` to a `expiries_at` on the restAPI side. But I wanted to keep the interference of the restAPI with the auth service as minimal as possible - that is why I decided to simply pass the token data as it was recieved by the auth instance.

**Review by MoSHad91 — APPROVED** [review]

**Comment by MoSHad91** [comment]
@antarcticrainforest 

I was in the middle of reviewing `freva-config` and didn't know where to write this. So i write here which seems more relevant to this:

Alright, if you would like to keep it as-is, I think `freva-config` repo has to be merged prior to `freva-nextgen` and beforehand some changes in this repo is needed. it seems you moved `keyclok` dir in freva-nextgen to `freva-config` and we don't need to have same files in both `freva-nextgen/dev-env/keycloak` and `freva-nextgen/dev-env/config/keycloak`. It needs some changes in https://github.com/FREVA-CLINT/freva-nextgen/blob/open-id-connect/dev-env/docker-compose.yaml as well
Same story for `redis-cmd.sh`.

But in principal, if keycloak is just designed for dev configuration, it is completely non-relevant to the definition/description of `freva-nextgen` repo.  Or you can at least put the whole `dev-utils.py` `keycloak` dev dir and so on in a same dev dir. 

> This repository holds all definitions for Docker images to create services that are needed to run Freva in production and development mode

also i think you have to move the https://github.com/FREVA-CLINT/freva-nextgen/blob/open-id-connect/dev-env/docker-compose.yaml to `freva-config` dev dir and also remove the `dev-env/key.py` from here as you made `dev-utils.py` in `freva-config`

**Comment by antarcticrainforest** [comment]
> @antarcticrainforest
> 
> I was in the middle of reviewing `freva-config` and didn't know where to write this. So i write here which seems more relevant to this:
> 
> Alright, if you would like to keep it as-is, I think `freva-config` repo has to be merged prior to `freva-nextgen` and beforehand some changes in this repo is needed. it seems you moved `keyclok` dir in freva-nextgen to `freva-config` and we don't need to have same files in both `freva-nextgen/dev-env/keycloak` and `freva-nextgen/dev-env/config/keycloak`. It needs some changes in https://github.com/FREVA-CLINT/freva-nextgen/blob/open-id-connect/dev-env/docker-compose.yaml as well Same story for `redis-cmd.sh`.
> 
> But in principal, if keycloak is just designed for dev configuration, it is completely non-relevant to the definition/description of `freva-nextgen` repo. Or you can at least put the whole `dev-utils.py` `keycloak` dev dir and so on in a same dev dir.
> 
> > This repository holds all definitions for Docker images to create services that are needed to run Freva in production and development mode
> 
> also i think you have to move the https://github.com/FREVA-CLINT/freva-nextgen/blob/open-id-connect/dev-env/docker-compose.yaml to `freva-config` dev dir and also remove the `dev-env/key.py` from here as you made `dev-utils.py` in `freva-config`

Hmm,

yes I will need to change the freva-nexgen, but obviously after this here has been merged.

it's not only used for production but also for development. This is why we used this repo as a submodule. In freva-web and freva. 

If you don't like the approach that keycloak and open ldap is on the same level as the other services I can move it to a dev folder to make in more clear that's dev. But imho I don't think that it harms if this  definitions aboard even if they are not used in production.

About the docker-compose files. I thought about this but ditched the idea because if you wanted to make the smallest change, like adding a new env var. You'd have to submit a PR here and wait for it to be merged.  This is too painful if you ask me. So I decided to leave the docker-compose files where they are. But I'll think about something.

**Comment by antarcticrainforest** [comment]
I've unified the token responses in the rest api so that it matches to token in the client. It also works with authlib.

File: freva-client/src/freva_client/auth.py
Comment (MoSHad91) [inline freva-client/src/freva_client/auth.py:57]: https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims:~:text=scope%3Dopenid%20profile%20email%20phone

File: freva-rest/src/freva_rest/auth.py
Comment (MoSHad91) [inline freva-rest/src/freva_rest/auth.py:130]: hmm, I'm not against of defining so many ORs for different `expires` keywords, but if you really would like to make it scaleable for the future, go for the approach that I mentioned in first review and defining a expiration_keys dictionary. 

https://github.com/FREVA-CLINT/freva-nextgen/pull/54#discussion_r1703859375

**Review by MoSHad91 — APPROVED** [review]
lgtm🚀

File: freva-rest/src/freva_rest/auth.py
Comment (antarcticrainforest) [reply freva-rest/src/freva_rest/auth.py:130]: That's not really true. The piece of code you posted, a) loops every time through keys and then b) checks for each entry if the key is within the in hash map (aka dict) that's. Under the hood that's another loop. And hence O(n^2). Retrieving items from a hash map is O(1) checking their presence O(n). So it's always better to access from a hashmap rather than checking the presence in a hashmap. Take a look at this:


```python
 def foo_get():
    lookup = {"expires_in": 200}
    return lookup.get("exp") or lookup.get("expires") or lookup.get("expires_at") or lookup.get("expires_in")

def foo_loop():
   lookup = {"expires_in": 200}
   keys = ["exp", "expires", "expires_at", "expires_in"]
   for key in keys:
      if key in lookup:
         return lookup[key]
         
%timeit foo_get()
460 ns ± 53.5 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

%timeit foo_loop()
726 ns ± 128 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

**Pull request merged**
